datashader/transfer_functions.py

```diff
@@ -163,3 +166,144 @@ def colorize(agg, color_key, how='cbrt', min_alpha=20):
+    return Image(out, dims=img.dims, coords=img.coords)
+
+
+@memoize
+def _build_spread_kernel(how):
+    """Build a spreading kernel for a given composite operator"""
+    op = composite_op_lookup[how]
+
+    @nb.jit(nopython=True, nogil=True)
+    def kernel(arr, mask, out):
+        M, N = arr.shape
+        w = mask.shape[0]
+        for y in range(M):
+            for x in range(N):
+                el = arr[y, x]
+                # Skip if data is transparent
+                if (el >> 24) & 255:
+                    for i in range(w):
+                        for j in range(w):
+                            # Skip if mask is False at this value
+                            if mask[i, j]:
+                                out[i + y, j + x] = op(el, out[i + y, j + x])
+    return kernel
+
+
+def _square_mask(px):
+    """Produce a square mask with sides of length ``2 * px + 1``"""
+    px = int(px)
+    w = 2 * px + 1
+    return np.ones((w, w), dtype='bool')
+
+
+def _circle_mask(r):
+    """Produce a circular mask with a diameter of ``2 * r + 1``"""
+    r = int(r)
+    w = 2 * r + 1
+    mask = np.zeros((w, w), dtype='bool')
+    _fill_circle(mask, r, r, r, True)
+    return mask
+
+
+_mask_lookup = {'square': _square_mask,
+                'circle': _circle_mask}
+
+
+@nb.jit(nopython=True, nogil=True)
+def _fill_circle(arr, x0, y0, r, val):
+    """General circle filling routine.
+
+    Low level, no bounds checking performed. Has potential to segfault if
+    misused.
+
+    Parameters
+    ----------
+    arr : 2d array
+    x0, y0 : int
+        Center coordinates of the circle.
+    r : int
+        Circle radius in pixels. Note that this excludes the center pixel - the
+        resulting circle will be ``2 * r + 1`` pixels in diameter.
+    val : scalar
+        Fill value
+    """
+    x = r
+    y = 0
+    do2 = 1 - x
+
+    # Special cases for r == 0 (fastpath) and r == 1. The normal algorithm will
+    # draw a 3x3 grid for this case, but a "+" looks a bit better in my opinion
+    if r == 0:
+        arr[y0, x0] = 1
+        return
+    elif r == 1:
+        _fill_span(arr, y0, x0 - 1, x0 + 1, val)
+        arr[y0 + 1, x0] = arr[y0 - 1, x0] = val
+        return
+    # Determine width
+    while y <= x:
+        y += 1
+        if do2 <= 0:
+            do2 += 2 * y + 1
+        else:
+            x -= 1
+            do2 += 2 * (y - x) + 1
+            break
+    width = y - 1
+    # Fill in spans
+    _fill_span(arr, y0 + r, x0 - width, x0 + width, val)
+    _fill_span(arr, y0 - r, x0 - width, x0 + width, val)
+    for row in range(-width, width + 1):
+        _fill_span(arr, y0 + row, x0 - r, x0 + r, val)
+    # Fill in remainder
+    while y < r:
+        _fill_span(arr, y0 + y, x0 - x, x0 + x, val)
+        _fill_span(arr, y0 - y, x0 - x, x0 + x, val)
+        y += 1
+        if do2 <= 0:
+            do2 += 2 * y + 1
+        else:
+            x -= 1
+            do2 += 2 * (y - x) + 1
+
```

Comment:
Just curious -- this looks like a pretty complicated implementation; what does it buy over something more like:

```python
>>> import numpy as np                                                                                                                                  
>>> def disk(x, y, radius):                                                                                                                             
      distance_from_origin = np.sqrt(x**2+y**2)                                                                                                         
      distance_outside_disk = distance_from_origin - radius                                                                                             
      return np.where(distance_from_origin - radius<=0,1.0,0.0)
>>> x,y = np.mgrid[-5:5, -5:5] * 0.1
>>> disk(x,y,0.2)
array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  1.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  1.,  1.,  1.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  1.,  1.,  1.,  1.,  1.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  1.,  1.,  1.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  1.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]])
```